rfc-drafts/must_not_await_lint.md

```diff
@@ -0,0 +1,101 @@
+
+The big reason for including a lint like this is because under the hood the compiler will automatically transform async fn into a state machine which can store locals. This process is invisible to users and will produce code that is different than what is in the actual rust file. Due to this it is important to inform users that their code may not do what they expect.
+
+# Guide-level explanation
+
+Provide a lint that can be attached to structs to let the compiler know that this struct can not be held across an await boundary.
+
+```rust
+    #[must_not_await]
+    struct MyStruct {}
+```
+
+This struct if held across an await boundary would cause a warning:
+
+```rust
+    async fn foo() {
+      let my_struct = MyStruct {};
+      my_async_op.await;
+      println!("{:?}", my_struct);
+    }
+```
+
+The compiler might output something along the lines of:
+
+TODO: Write a better error message.
+```
+warning: Holding `MyStruct` across the await bound on line 3 might cause side effects.
+```
+
+Example use cases for this lint:
+
+- `MutexGuard` holding this across a yield boundary in a single threaded executor could cause deadlocks. In a multi-threaded runtime the resulting future would become `!Send` which will stop the user from spawning this future and causing issues. But in a single threaded runtime which accepts `!Send` futures deadlocks could happen.
+
+- The same applies to other such synchronization primitives such as locks from `parking-lot`.
+
+- `tracing::Span` has the ability to enter the span via the `tracing::span::Entered` guard. While entering a span is totally normal, during an async fn the span only needs to be entered once before the `.await` call, which might potentially yield the execution.
+
+- Any RAII guard might possibly create unintended behavior if held across an await boundary.
+
+This lint will enable the compiler to warn the user that the generated MIR could produce unforeseen side effects. Some examples of this are:
+
+- [`std::sync::MutexGuard`](https://doc.rust-lang.org/std/sync/struct.MutexGuard.html)
+- [`tracing::span::Entered`](https://docs.rs/tracing/0.1.15/tracing/span/struct.Entered.html)
+
+This will be a best effort lint to signal to the user about unintended side-effects of using certain types across an await boundary.
+
+# Reference-level explanation
+
+Going throuogh the prior are we see two systems currently which provide simailar/semantically similar behavior:
```

Comment:
This seems more like it belongs in either the motivation or the rationale and alternatives section, to me at least.